File: permission/src/main/java/com/banzhi/permission/PermissionInit.java
package com.banzhi.permission;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Application;
import android.content.Context;

import java.lang.reflect.InvocationTargetException;

/**
 * <pre>
 * @author : No.1
 * @time : 2019/5/28.
 * @desciption :
 * @version :
 * </pre>
 */

public class PermissionInit {
    private static Application sApplication;
    private static final PermissionActivityLifecycle lifecycle = new PermissionActivityLifecycle();

    public static void init(Context context) {
        if (context == null) {
            init(getApplicationByReflect());
        } else {
            init((Application) context.getApplicationContext());
        }
    }

    public static void init(Application app) {
        if (sApplication == null) {
            if (app == null) {
                sApplication = getApplicationByReflect();
            } else {
                sApplication = app;
            }
            sApplication.registerActivityLifecycleCallbacks(lifecycle);
        } else {
            if (app != null && app.getClass() != sApplication.getClass()) {
                sApplication.unregisterActivityLifecycleCallbacks(lifecycle);
                lifecycle.clear();
                sApplication = app;
                sApplication.registerActivityLifecycleCallbacks(lifecycle);
            }
        }
    }

    private static Application getApplicationByReflect() {
        try {
            @SuppressLint("PrivateApi")
            Class<?> activityThread = Class.forName("android.app.ActivityThread");
            Object thread = activityThread.getMethod("currentActivityThread").invoke(null);
            Object app = activityThread.getMethod("getApplication").invoke(thread);
            if (app == null) {
                throw new NullPointerException("u should init first");
            }
            return (Application) app;
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        throw new NullPointerException("u should init first");
    }

    public static Activity getTopActivity() {
        return lifecycle.getTopActivity();
    }
}

File: permission_kt/src/main/java/com/banzhi/permission_kt/PermissionInit.kt
package com.banzhi.permission_kt

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application
import android.content.Context
import android.util.Log
import androidx.fragment.app.FragmentActivity
import java.lang.reflect.InvocationTargetException

/**
 *<pre>
 * @author : No.1
 * @time : 2019/8/7.
 * @desciption :
 * @version :
 *</pre>
 */
class PermissionInit {

    companion object {
        private var sApplication: Application? = null
        private val lifecycle = PermissionActivityLifecycle()

        fun init(context: Context?) {
            if (context == null) {
                init(getApplicationByReflect())
            } else {
                init(context.applicationContext as Application)
            }
        }

        fun init(app: Application?) {
            Log.e("init","PermissionInit")
            if (sApplication == null) {
                sApplication
                if (app == null) {
                    sApplication = getApplicationByReflect()
                } else {
                    sApplication = app
                }
                sApplication!!.registerActivityLifecycleCallbacks(lifecycle)
            } else {
                if (app != null && app.javaClass != sApplication!!.javaClass) {
                    sApplication!!.unregisterActivityLifecycleCallbacks(lifecycle)
                    lifecycle.clear()
                    sApplication = app
                    sApplication!!.registerActivityLifecycleCallbacks(lifecycle)
                }
            }
        }

        private fun getApplicationByReflect(): Application {
            try {
                @SuppressLint("PrivateApi")
                val activityThread = Class.forName("android.app.ActivityThread")
                val thread = activityThread.getMethod("currentActivityThread").invoke(null)
                val app = activityThread.getMethod("getApplication").invoke(thread)
                        ?: throw NullPointerException("u should init first")
                return app as Application
            } catch (e: NoSuchMethodException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            } catch (e: InvocationTargetException) {
                e.printStackTrace()
            } catch (e: ClassNotFoundException) {
                e.printStackTrace()
            }

            throw NullPointerException("u should init first")
        }

        fun getTopActivity(): Activity? {

            return lifecycle.getTopActivity()
        }
    }
}